external-crates/move/crates/move-compiler/src/linters/shift_overflow.rs

```diff
@@ -0,0 +1,108 @@
+//! Detect potential overflow scenarios where the number of bits being shifted exceeds the bit width of
+//! the variable being shifted, which could lead to unintended behavior or loss of data. If such a
+//! potential overflow is detected, a warning is generated to alert the developer.
+use crate::{
+    diag,
+    diagnostics::{
+        codes::{custom, DiagnosticInfo, Severity},
+        WarningFilters,
+    },
+    expansion::ast::Value_,
+    naming::ast::{BuiltinTypeName_, Type_},
+    parser::ast::BinOp_,
+    shared::{program_info::TypingProgramInfo, CompilationEnv},
+    typing::{
+        ast::{self as T, UnannotatedExp_},
+        visitor::{TypingVisitorConstructor, TypingVisitorContext},
+    },
+};
+use move_ir_types::location::Loc;
+
+use super::{LinterDiagCategory, LINTER_DEFAULT_DIAG_CODE, LINT_WARNING_PREFIX};
+
+const SHIFT_OPERATION_OVERFLOW_DIAG: DiagnosticInfo = custom(
+    LINT_WARNING_PREFIX,
+    Severity::Warning,
+    LinterDiagCategory::ShiftOperationOverflow as u8,
+    LINTER_DEFAULT_DIAG_CODE,
+    "Potential overflow detected. The number of bits being shifted exceeds the bit width of the variable being shifted.",
+);
+
+pub struct ShiftOperationOverflow;
+
+pub struct Context<'a> {
+    env: &'a mut CompilationEnv,
+}
+
+impl TypingVisitorConstructor for ShiftOperationOverflow {
+    type Context<'a> = Context<'a>;
+
+    fn context<'a>(
+        env: &'a mut CompilationEnv,
+        _program_info: &'a TypingProgramInfo,
+        _program: &T::Program_,
+    ) -> Self::Context<'a> {
+        Context { env }
+    }
+}
+
+impl TypingVisitorContext for Context<'_> {
+    fn visit_exp_custom(&mut self, exp: &mut T::Exp) -> bool {
```

Comment:
It might be worth moving this to the CFGIR, since we will get constant folding and optimizations run beforehand

Reply:
@cgswords I wonder if we should just add these as normal compiler warnings? We are basically 99% of the way there in external-crates/move/crates/move-compiler/src/cfgir/optimize/constant_fold.rs